src/cclib/io/moldenwriter.py

```diff
@@ -8,31 +8,53 @@
 """A writer for MOLDEN format files."""
 
 import os.path
+import math
 
-from . import filewriter
+from decimal import Decimal
 from cclib.parser import utils
+from . import filewriter
+
+
+def round_molden(num, p=6):
+    """Molden style number rounding in [Atoms] section."""
+    # Digit at pth position after dot.
+    p_digit = math.floor(abs(num) * 10 ** p) % 10
+    # If the 6th digit after dot is greater than 5, but is not 7,
+    # round the number upto 6th place.
+    # Else truncate at 6th digit after dot.
+    if p_digit > 5 and p_digit != 7:
+        return round(num, p)
+    if num >= 0:
+        return math.floor(num * 10 ** p) / 10 ** p
+    else:
+        return math.ceil(num * 10 ** p) / 10 ** p
 
 
 class MOLDEN(filewriter.Writer):
     """A writer for MOLDEN files."""
 
+    required_attrs = ['atomcoords', 'atomnos', 'natom']
+
     def _title(self, path):
         """Return filename without extension to be used as title."""
         title = os.path.basename(os.path.splitext(path)[0])
         return title
 
     def _coords_from_ccdata(self, index):
         """Create [Atoms] section using geometry at the given index."""
-
+        elements = [self.pt.element[Z] for Z in self.ccdata.atomnos]
         atomcoords = self.ccdata.atomcoords[index]
         atomnos = self.ccdata.atomnos
         nos = range(self.ccdata.natom)
 
         # element_name number atomic_number x y z
         atom_template = '{:2s} {:5d} {:2d} {:12.6f} {:12.6f} {:12.6f}'
         lines = []
-        for element, no, atomno, (x, y, z) in zip(self.elements, nos, atomnos,
+        for element, no, atomno, (x, y, z) in zip(elements, nos, atomnos,
                                                   atomcoords):
+            x = round_molden(x)
```

Comment:
You can avoid x, y and z in the loop definition here by doing `x, y, z = [round_molden(c) for c in coord]` or functioanlly `x, y, z = map(round_molden, coord)`.